numpy/core/src/umath/ufunc_type_resolution.c

```diff
@@ -1508,6 +1507,22 @@ ufunc_loop_matches(PyUFuncObject *self,
             tmp = dtypes[i];
             Py_INCREF(tmp);
         }
+        else if (types[i] == PyArray_TYPE(op[i])) {
+            /*
+             * Don't bother doing casting checks if the type numbers exactly
+             * match.
```

Comment:
Is the worry that we might have something like a binary function taking two datetime/timedelta inputs that depends on the unit metadata matching exactly?  I'd naively expect that it'd be that function's responsibility to validate metadata-specifc constraints.